File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/submodel/metamodel/api/submodelelement/entity/EntityType.java
package org.eclipse.basyx.submodel.metamodel.api.submodelelement.entity;

/**
 * Enumeration for denoting whether an entity is a self-managed entity or a
 * comanaged entity.
 * 
 * @author schnicke
 *
 */
public enum EntityType {
	/**
	 * For co-managed entities there is no separate AAS. Co-managed entities need to
	 * be part of a self-managed entity.
	 */
	COMANAGEDENTITY("CoManagedEntity"),
	
	/**
	 * Self-Managed Entities have their own AAS but can be part of the bill of
	 * material of a composite self-managed entity. The asset of an I4.0 Component
	 * is a self-managed entity per definition.
	 */
	SELFMANAGEDENTITY("SelfManagedEntity");
	
	private String name;

	private EntityType(String name) {
		this.name = name;
	}

	@Override
	public String toString() {
		return name;
	}

	public static EntityType fromString(String str) {
		if (str == null || str.isEmpty()) {
			return null;
		}

		if (str.equals(EntityType.COMANAGEDENTITY.toString())) {
			return COMANAGEDENTITY;
		} else if (str.equals(EntityType.SELFMANAGEDENTITY.toString())) {
			return SELFMANAGEDENTITY;
		}
		throw new RuntimeException("Unknown EntityType: " + str);
	}
}

File: sdks/c++/basys.sdk.cc/src/submodel/submodel/enumerations/EntityType.cpp
#include <BaSyx/submodel/enumerations/EntityType.h>

#include <array>
#include <algorithm>
#include <memory>
#include <string>

using namespace basyx::submodel;

using enum_pair_t = std::pair<const char*, EntityType>;

static const std::array<enum_pair_t, 2> string_to_enum = 
{
    std::make_pair("CoManagedEntity",  EntityType::CoManagedEntity),
    std::make_pair("SelfManagedEntity", EntityType::SelfManagedEntity),
};

EntityType EntityType_::from_string(const std::string & name)
{
    auto pair = std::find_if(string_to_enum.begin(), string_to_enum.end(), 
		[&name](const enum_pair_t & pair) {
			return !name.compare(pair.first);
	});

    return pair->second;
}

const char * EntityType_::to_string(EntityType value)
{
    auto pair = std::find_if(string_to_enum.begin(), string_to_enum.end(), 
		[value](const enum_pair_t & pair) {
			return value == pair.second;
	});

    return pair->first;
}

